Editor: Move printing TinyMCE scripts into an action

sarayourfriend opened the pull request:
This is a draft PR to start addressing the issue in the ticket. Moving this logic into an action allows plugins to override how/when TinyMCE is loaded. Specifically, Gutenberg needs to be able to prevent TinyMCE scripts from being written into the page when TinyMCE should be loaded asynchronously when it is actually needed.

This is the first time I've made changes in WordPress core and I had a hard time deciding where to put the `wp_print_tinymce_scripts` function (or what to name it, for that matter).

I'm also not even sure that this is exactly the correct approach for the issue described in the ticket, so any and all feedback is super welcome!

Trac ticket: https://core.trac.wordpress.org/ticket/49964

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

sarayourfriend committed:
Editor: Move printing TinyMCE scripts into an action

Moving this logic into an action allows plugins to override
how/when TinyMCE is loaded. Specifically, Gutenberg needs to
be able to prevent TinyMCE scripts from being written into
the page when TinyMCE should be loaded asynchronously when it
is actually needed.

saramarcondes commented on src/wp-includes/class-wp-editor.php:
```diff
@@ -1556,9 +1556,7 @@ public static function print_tinymce_scripts() {
-		wp_print_scripts( array( 'wp-tinymce' ) );
-
-		echo "<script type='text/javascript'>\n" . self::wp_mce_translation() . "</script>\n";
+		do_action( 'print_tinymce_scripts' );
```
I wasn't sure how to add tests for this change, would love some pointers on that.

I tried testing around `did_action( 'print_tinymce_scriots' )` but it seems like each test isn't running with a brand new WP so `did_action` returns `true` before doing anything inside the test body. Is there a way to observe whether this action is performed when I call `_WP_Editors::print_tinymce_scripts()` directly? Is this a worthwhile test?

nb replied:
Hm, @saramarcondes, I can't reproduce this `did_action` behavior in tests. It clears up for me. Here is my attempt to reduce the issue: 

```php
<?php

class Test_Did_Action extends WP_UnitTestCase {
	public function test_did_action_should_be_two_if_action_fired_twice() {
		do_action( 'wink' );
		do_action( 'wink' );
		$this->assertEquals( 2, did_action( 'wink' ) );
	}

	public function test_did_action_should_be_one_if_action_fired_once() {
		do_action( 'wink' );
		$this->assertEquals( 1, did_action( 'wink' ) );
	}

	public function test_did_action_should_be_zero_if_no_action_fired() {
		$this->assertEquals( 0, did_action( 'wink' ) );
	}
}
```

If WordPress doesn't clean its globals (this is where `do_action`/`did_action` [keep information](https://github.com/WordPress/wordpress-develop/blob/3659b767f1f8a4d3ca08f6d8310963b9d3a73b08/src/wp-includes/plugin.php#L493-L501)) at least one of the cases would have failed (if we assume they could have been executed in any order).

If something else is firing the action (code outside of our control), maybe we can test whether the difference of the return value of `did_action` before and after our code is what we want? Zero if we don't expect our code to fire the action, or one if we expect it to fire once?

saramarcondes replied:
Ah! I believe the issue is that I incorrectly assumed `did_action` returns `boolean` and `assertFalse( 0 )` isn't going to pass 😀

nb replied:
Haha, I can't blame you – I wish anything starting with `did_` only returned a boolean, but 🤷‍♂️

sarayourfriend committed:
Script Loader: Add method get_url for getting the full URI of a script handle
Based on https://github.com/WordPress/gutenberg/pull/21244 but pulled into
Core so that this functionality can be used when retriving the URIs for
TinyMCE's dependencies when it is being asynchronously loaded.

sarayourfriend committed:
Editor: Move meta box processing before TinyMCE script enqueuing

This is necessary in order to support lazily loading TinyMCE for the
classic block. Because some custom meta boxes may depend on TinyMCE,
if they are present we must pre-load TinyMCE as normal. To be able to
support making that decision at runtime, we need to have already
processed and initialized the `$wp_meta_boxes` global when we call
to wp_enqueue_editor, which the Gutenberg plugin hooks into through
a new action also added as part of this patch to programatically
decide whether or not TinyMCE needs to be pre-loaded.
